### PR TITLE
install: allow specifying external CA template

### PR DESCRIPTION
**Pull request by frasertweedale**
This PR allow an admin to specify an AD-CS target certificate template
by name or OID, via the new option --external-ca-profile.
The profile may be specified by name (string) or OID + version + optional minor version.

https://pagure.io/freeipa/issue/6858

The approach is:

1. preliminary refactor to the IPAOptionParser to allow easily specifying a custom
data constructor (this is used for the data type that holds the template specifier).

2. refactor to reduce duplication of external CA type enum values.

3. the main thing:
   - add data type for template specifier
   - add ipa-server-install `--external-ca-profile` CLI option
   - update CA installation to add the appropriate *pkispawn* config based on the template
    specifier
   - update *ipa-server-install* man page

4. add the `external-ca-profile` option to *ipa-ca-install* and update man page.

5. update ipa-cacert-manage to be able to specify the new options

6. tests, minor fixes and cleanups

**NOTE FOR TESTERS**

*python-cryptography* had a bug parsing long OIDs.  AD-CS creates and uses OIDs long enough to trigger the bug as a matter of course.  The bug is fixed as of v1.9.  **f26** and **f27**
include a fixed version.

certmonger-0.79.5 is required.  It is available for **f27** and a build for f26 is available in
the **freeipa-master COPR**.

**HOW TO TEST**

1. Install AD-CS in a Windows machine and create a custom profile by copying the *SubCA*
profile.

2. Two-step external CA ipa-server-install:

```
$ ipa-server-install \
    --external-ca --external-ca-type=ms-cs \
    --external-ca-profile=1.3.6.1.4.1.311.21.8.8950086.10656446.2706058.12775672.480128.147.7130143.4405632:1
```

(Use the actual OID of the custom profile).  If everything works, hooray!

3. Start over with ca-less deployment.  Then add CA via ``ipa-ca-install --external-ca-... # as before``.
  If everything works, hooray.

4. test that `ipa-cacert-manage renew` with the new options results in the template extensions being included in the new CSR.

### CODE REVIEW

**Comment by frasertweedale** [comment]
COPR build: https://copr.fedorainfracloud.org/coprs/ftweedal/freeipa-adcs-template

**Comment by frasertweedale** [comment]
Blog post explaining how to set up AD CS and configure a custom sub-CA template, for testing this feature: https://blog-ftweedal.rhcloud.com/2017/08/installing-freeipa-with-an-active-directory-subordinate-ca/


Also, I just remembered I will also need to modify `ipa-cacert-manage renew` to allow specifying
the template during renewal.  So either I will update this PR or - if it has been reviewed and merged by the time I implement this - create a follow-up PR with that change.

**Comment by frasertweedale** [comment]
I've added some more commits for the `ipa-cacert-manage renew` scenario.
These require a new build of certmonger.

Certmonger PR: https://pagure.io/certmonger/pull-request/81
COPR: https://copr.fedorainfracloud.org/coprs/ftweedal/certmonger-v2template

After the Certmonger changes get merged and a new release made, I'll
bump the min version of certmonger in the spec file.

**Comment by frasertweedale** [comment]
I updated the blog post, adding a section about renewal and the certmonger changes:
http://blog-ftweedal.rhcloud.com/2017/08/installing-freeipa-with-an-active-directory-subordinate-ca/#renewing-the-certificate

File: ipaserver/install/cainstance.py [hunk 1]
```diff
@@ -565,12 +578,18 @@ def __spawn_instance(self):
...
 
-            if self.ca_type == 'ms-cs':
+            if self.ca_type == ExternalCAType.MS_CS.value:
                 # Include MS template name extension in the CSR
+                template = self.external_ca_profile
+                if template is None:
+                    # default template name
+                    template = MSCSTemplateV1(u"SubCA")
+
+                ext_data = binascii.hexlify(template.get_ext_data())
```
[the remark this answers is not in the document]
Comment (frasertweedale) [reply ipaserver/install/cainstance.py]: Thanks @flo-renaud, thanks for picking this up; this PR has been hanging around a while :D

**Comment by frasertweedale** [comment]
@flo-renaud I think I have resolved the py3 issues.

The only failing CI job is lint, which is failing with
```
************* Module ipatests.test_ipaserver.test_install.test_cainstance

ipatests/test_ipaserver/test_install/test_cainstance.py:7: [E0401(import-error), ] Unable to import 'six.moves.configparser')
```

That module definitely exists so I'm not sure what's going on there.

**Comment by flo-renaud** [comment]
Hi @frasertweedale 
thank you for the updated PR. I found an issue in the scenario "switch from ipa self-signed ca to windows-signed ca":
- install IPA with self-signed CA
- run ipa-cacert-manage install to add windows CA, run ipa-certupdate
- run ipa-cacert-manage renew --external-ca --external-ca-type=ms-cs --external-ca-profile=<OID>:1 to generate the csr (or use --external-ca-profile=MySubCA)
- run ipa-cacert-manage renew --external-cert-file ipaca.cert --external-cert-file winca.cert
Importing the renewed CA certificate, please wait
Subject name encoding mismatch (visit http://www.freeipa.org/page/Troubleshooting for troubleshooting guide)
The ipa-cacert-manage command failed.

Strangely, old_cert.subject == new_cert.subject but old_cert.subject_bytes != new_cert.subject_bytes.

The reverse scenario (install with windows-signed ca, switch to self-signed) also fails:
sudo ipa-cacert-manage renew --self-signed
Renewing CA certificate, please wait
Error resubmitting certmonger request '20170927084855', please check the request manually
The ipa-cacert-manage command failed.

Other scenarios are working:
- ipa-server-install with windows-signed CA (with name or OID), renew with windows CA
- ipa-server-install CA-less, ipa-ca-install with windows-signed CA

**Comment by frasertweedale** [comment]
@flo-renaud aha, excellent pickup!  This might be caused by default configuration of AD-CS.  AD-CS
can change the encoding of the Subject DN under some circumstances, making it different from what
is in the CSR.  Let me follow up tomorrow to confirm if it can be addressed by a change in AD-CS
configuration, or whether there is something wrong with the CSR itself.

Thank you for the thorough testing!

**Comment by frasertweedale** [comment]
@flo-renaud could you please provide the following info:

- Precise encoding of subject name in self-signed IPA CA certificate.
- Precise encoding of subject name in CSR generated by `ipa-cacert-manage renew`
- On your AD machine, the output of command `certutil -getreg ca\forceteletex`

Feel free to just paste the whole PEM-encoded cert/CSR.

**Comment by flo-renaud** [comment]
Hi @frasertweedale,

My test machine is Windows 2008:
```
C:\Users\Administrator>certutil -getreg ca\forceteletex
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\CertSvc\Configuration\dom-058-183-VM-058-183-CA\ForceTeletex:

  ForceTeletex REG_DWORD = 12 (18)
    ENUM_TELETEX_AUTO -- 2
    ENUM_TELETEX_UTF8 -- 10 (16)
CertUtil: -getreg command completed successfully.

```

The cert/csr were sent to you in private.

**Comment by frasertweedale** [comment]
Thanks @flo-renaud.  So as I suspected, the problem is that although the Subject DN in the
CSR uses `UTF8String` encoding, the CA produces a cert with the Subject DN attributes encoded
using `PrintableString`.  The problem is not in FreeIPA but is a problem in the AD-CS configuration.

So, how to make AD-CS respect the attribute encoding?  Execute the following commands:

    certutil -setreg ca\forceteletex +0x20
    net stop "certificate services"
    net start "certificate services"

See also: https://support.microsoft.com/en-us/help/888180/how-to-make-a-stand-alone-certification-authority-that-is-running-wind.

After this configuration change, hopefully you will find that the `ipa-cacert-manage renew` operations succeed.

**Comment by flo-renaud** [comment]
Hi @frasertweedale,
thank you for the pointer. After applying the modification to my Windows CS, the scenario 'switch from self-signed to windows-signed' is succeeding. But I still have issues when doing the reverse, ie. switch from Win-signed to self-signed.

**Comment by frasertweedale** [comment]
@flo-renaud I'm fighting regressions in Dogtag `master`, but in the meantime could you please provide more precise detail on what the failure was in the externally-signed to self-signed switch?

**Comment by flo-renaud** [comment]
@frasertweedale 
don't worry for this issue, it is not linked to your PR. The issue also happens on the master branch when switching from externally signed CA to self-signed CA:

> $ sudo ipa-cacert-manage renew --self-signed
> Renewing CA certificate, please wait
> Error resubmitting certmonger request '20170929123437', please check the request manually
> The ipa-cacert-manage command failed.
> $ sudo getcert list -i 20170929123437
> Number of certificates and requests being tracked: 9.
> Request ID '20170929123437':
> 	status: MONITORING
> 	ca-error: Updated certificate not available

Issue https://pagure.io/freeipa/issue/7173 opened to track this

**Comment by frasertweedale** [comment]
@flo-renaud thank you for reviewing!  I know this was a complex enhancement to test.

**Comment by pvomacka** [comment]
master:

* c5afee964eee0cdf81a4f22fd78a6838a3da7537 cli: simplify parsing of arbitrary types
* 1699cff3501519982d896bd5d5add233d12b0f00 Remove duplicate references to external CA type
* b4365e3a7fa2fb1d4e7ffc41f21c23e3d369ffd6 install: allow specifying external CA template
* fc7c684b122cbcf1ff31c0ee45551388abb588fc ipa-ca-install: add --external-ca-profile option
* 2207dc5c172710471f3c7c77242cb2ba1fcfa779 certmonger: refactor 'resubmit_request' and 'modify'
* 560ee3c0b512cbb8cdc4099a81204e745a515f7c certmonger: add support for MS V2 template
* 29f4ec865b2d0654e7e46c8a089a04f23ea9a00c ipa-cacert-manage: support MS V2 template extension
* d43cf35cca4e7b84801bbbd5b1bb910d6a5e453a Add tests for external CA profile specifiers
* 49c0a7b4d4ec5ddef7f9648be72ccacb15c28840 ipa-cacert-manage: handle alternative tracking request CA name
* 75a2eda85d7e9769f612733a6ade5719b1511c09 ipa-cacert-manage: avoid some duplicate string definitions

**Pull request closed without merging**